exec/cnex/lib/posix.c

```diff
@@ -0,0 +1,781 @@
+
+    push(exec->stack, wrap(pathconf(path, number_to_sint32(name))));
+    free(path);
+}
+
+void posix_pipe(struct tagTExecutor *exec)
+{
+    int fds[2];
+    int r = pipe(fds);
+    if (r < 0) {
+        push(exec->stack, wrap(r));
+        return;
+    }
+    Cell *rfd = cell_fromNumber(number_from_sint32(fds[0]));
+    Cell *wfd = cell_fromNumber(number_from_sint32(fds[1]));
+
+    push(exec->stack, wrap(r));
+    push(exec->stack, wfd);
+    push(exec->stack, rfd);
+}
+
+void posix_pread(struct tagTExecutor *exec)
+{
+    Number offset = top(exec->stack)->number; pop(exec->stack);
+    Number nbyte = top(exec->stack)->number; pop(exec->stack);
+    Number fildes = top(exec->stack)->number; pop(exec->stack);
+
+    size_t n = number_to_uint64(nbyte);
+    Cell *buf = cell_createStringCell(n);
+    buf->type = cBytes;
+
+    push(exec->stack, wrap(pread(number_to_sint32(fildes), buf->string->data, n, number_to_uint64(offset))));
+    push(exec->stack, buf);
+}
+
+void posix_pwrite(struct tagTExecutor *exec)
+{
+    Number offset = top(exec->stack)->number; pop(exec->stack);
+    Number nbyte = top(exec->stack)->number; pop(exec->stack);
+    TString *buf = top(exec->stack)->string;
+    Number fildes = peek(exec->stack, 1)->number;
+
+    Cell *r = wrap(pwrite(number_to_sint32(fildes), buf->data, number_to_uint64(nbyte), number_to_uint64(offset)));
+    pop(exec->stack);
+    pop(exec->stack);
+
+    push(exec->stack, r);
+}
+
+void posix_read(struct tagTExecutor *exec)
+{
+    Number nbyte = top(exec->stack)->number; pop(exec->stack);
+    Number fildes = top(exec->stack)->number; pop(exec->stack);
+
+    size_t n = number_to_uint64(nbyte);
+    Cell *buf = cell_createStringCell(n); buf->type = cBytes;
+    Cell *r =  wrap(read(number_to_sint32(fildes), buf->string->data, n));
+
+    push(exec->stack, buf);
+    push(exec->stack, r);
+}
+
+void posix_readlink(struct tagTExecutor *exec)
+{
+    Number bufsize = top(exec->stack)->number; pop(exec->stack);
+    char *path = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    size_t n = number_to_uint64(bufsize);
+    Cell *buf = cell_createStringCell(n); buf->type= cBytes;
+    Cell *r = wrap(readlink(path, buf->string->data, n));
+
+    push(exec->stack, buf);
+    push(exec->stack, r);
+    free(path);
+}
+
+#if HAVE_AT_FUNCTIONS
+void posix_readlinkat(TExecutor *exec)
+{
+    Number bufsize = top(exec->stack)->number; pop(exec->stack);
+    char *path = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Number fd = top(exec->stack)->number; pop(exec->stack);
+
+    size_t n = number_to_uint64(bufsize);
+    Cell *buf = cell_createStringCell(n); buf->type = cBytes;
+    Cell *r = wrap(readlinkat(number_to_sint32(fd), path, buf->string->data, n));
+
+    push(exec->stack, buf);
+    push(exec->stack, r);
+}
+#endif
+
+void posix_rename(struct tagTExecutor *exec)
+{
+    char *new_ = string_asCString(top(exec->stack)->string); pop(exec->stack);
```

Comment:
Why do these local variables have the weird trailing underscores?

Reply:
I had modeled my code off posix.cpp; which might not have been the smartest thing to go, since we're in very different language universes.  Fixed.